[Product CRUD] Product crud related products

mikejolley opened the pull request:
Removes related product functions from classes to dedicated functions.

mikejolley committed:
Add related product functions and deprecate those in class.

mikejolley committed:
Merge branch 'product-crud' into product-crud-related-products

mikejolley committed:
No need to exclude ID

justinshreve commented on includes/wc-product-functions.php:
```diff
@@ -751,3 +751,123 @@ function wc_get_product_visibility_options() {
+	global $wpdb;
+
+	$product_id     = absint( $product_id );
+	$exclude_ids    = array_map( 'absint', array_merge( array( 0, $product_id ), $exclude_ids ) );
```
I think the `absint` sanitization should be inside `wc_get_related_products_query` (or in both places). Now that `wc_get_related_products_query` is a public global function, instead of a private function, people might use it directly to generate a query. So`$exclude_ids`, `$cats_array`, and `$tags_array` should be escaped as late as possible I think to prevent someone from making a mistake.

justinshreve commented on includes/wc-product-functions.php:
```diff
@@ -751,3 +751,123 @@ function wc_get_product_visibility_options() {
+		set_transient( $transient_name, $related_posts, DAY_IN_SECONDS );
+	}
+
+	// Randomise the results
```
We don't need some of the extra obvious comments here like "randomise the results", "limit the results", or "get the posts".

claudiosanches committed:
Fixed coding standards and improved the docblocks

claudiosanches committed:
Get cached terms from wc_get_related_terms()

claudiosanches committed:
Fixed wrong variable in wc_get_related_terms

claudiosanches committed:
Use count() instead of sizeof()

claudiosmweb commented:
@mikejolley I made some fixes based on WP CodeSniffer and fixed the `wc_get_related_terms()` function.

emdashcodes reviewed (commented):
PR looks pretty good. Just left a comment regarding comments, and about maybe moving some sanitization into the query function.

justinshreve commented on includes/wc-product-functions.php:
```diff
@@ -751,3 +751,133 @@ function wc_get_product_visibility_options() {
+		set_transient( $transient_name, $related_posts, DAY_IN_SECONDS );
+	}
+
+	// Randomise the results.
```
We don't need some of the extra obvious comments here like "randomise the results", "limit the results", or "get the posts".

claudiosmweb commented:
With this Pull Request we can make the follow function more simple:

```php
 function wc_get_product_cat_ids( $product_id ) {
 	$product_cats = wp_get_post_terms( $product_id, 'product_cat', array( "fields" => "ids" ) );
 
 	foreach ( $product_cats as $product_cat ) {
 		$product_cats = array_merge( $product_cats, get_ancestors( $product_cat, 'product_cat' ) );
 	}
 
 	return $product_cats;
}
```

It's possible to change `$product_cats = wp_get_post_terms( $product_id, 'product_cat', array( "fields" => "ids" ) );` to `$product_cats = wc_get_related_terms( $product_id, 'product_cat' );`.

mikejolley committed:
Sanitize ids later

mikejolley committed:
Remove unneeded comments

mikejolley commented:
@claudiosmweb Thinking about it, the naming of that function is bad. It's not getting related terms, it's getting the actual terms. Maybe it's possible to remove entirely and use the cached function you mentioned earlier.

claudiosmweb commented:
@mikejolley yes, `wc_get_related_terms()` works as `wc_get_product_terms()`.

mikejolley commented:
@claudiosmweb Does not need a filter either - get_the_terms has one.

mikejolley committed:
wc_get_product_term_ids instead of related wording and use in other places.

get_the_terms is used here and also handles caching, something
wp_get_post_terms does not.

mikejolley committed:
Bump template version

mikejolley commented:
@justinshreve ready to be re-reviewed and merged.

justinshreve commented on includes/wc-product-functions.php:
```diff
@@ -751,3 +751,116 @@ function wc_get_product_visibility_options() {
+ *
+ * @since  2.7.0
+ * @param  int    $product_id Product ID.
+ * @param  string $taxonomy   Taxonomy slug.
```
Incorrect comment

justinshreve commented on templates/single-product/related.php:
```diff
@@ -26,7 +26,7 @@
 }
 
-if ( ! $related = $product->get_related( $posts_per_page ) ) {
+if ( ! $related = wc_get_related_products( $product->get_id(), $posts_per_page, $product->get_upsells() ) ) {
```
This will throw a deprecation message. We renamed this to `get_upsell_ids`.

emdashcodes reviewed (approved):
Looks good, just pointed out one issue. I think it's good to merge once it's fixed 👍.

mikejolley committed:
Use get_upsell_ids
